amr-wind/physics/ConvectingTaylorVortex.cpp

```diff
@@ -0,0 +1,164 @@
+    , m_density(sim.repo().get_field("density"))
+{
+    amrex::ParmParse pp("CTV");
+    pp.query("density", m_rho);
+    pp.query("u0", m_u0);
+    pp.query("v0", m_v0);
+    {
+        amrex::Real nu;
+        amrex::ParmParse pp("transport");
+        pp.query("viscosity", nu);
+        m_omega = utils::pi() * nu;
+    }
+    if (amrex::ParallelDescriptor::IOProcessor()) {
+        std::ofstream f;
+        f.open(m_output_fname.c_str());
+        f << std::setw(m_w) << "time" << std::setw(m_w) << "L2_u"
+          << std::setw(m_w) << "L2_v" << std::endl;
+        f.close();
+    }
+}
+
+/** Initialize the velocity and density fields at the beginning of the
+ *  simulation.
+ */
+void ConvectingTaylorVortex::initialize_fields(
+    int level, const amrex::Geometry& geom)
+{
+    using namespace utils;
+
+    auto& velocity = m_velocity(level);
+    auto& density = m_density(level);
+
+    density.setVal(m_rho);
+
+    for (amrex::MFIter mfi(velocity); mfi.isValid(); ++mfi) {
+        const auto& vbx = mfi.validbox();
+
+        const auto& dx = geom.CellSizeArray();
+        const auto& problo = geom.ProbLoArray();
+        auto vel = velocity.array(mfi);
+
+        amrex::ParallelFor(
+            vbx, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+                const amrex::Real x = problo[0] + (i + 0.5) * dx[0];
+                const amrex::Real y = problo[1] + (j + 0.5) * dx[1];
+                vel(i, j, k, 0) = u_exact(m_u0, m_v0, m_omega, x, y, 0.0);
+                vel(i, j, k, 1) = v_exact(m_u0, m_v0, m_omega, x, y, 0.0);
+                vel(i, j, k, 2) = 0.0;
+            });
+    }
+}
+
+amrex::Real ConvectingTaylorVortex::compute_error(
+    const int comp, const Field& field, amr_wind::ctv::FuncDef f)
+{
+    amrex::Real error = 0.0;
+    const amrex::Real time = m_time.current_time();
+
+    const int nlevels = m_repo.num_active_levels();
+    for (int lev = 0; lev < nlevels; ++lev) {
+
+        amrex::iMultiFab level_mask;
+        if (lev < nlevels - 1) {
+            level_mask = makeFineMask(
+                m_mesh.boxArray(lev), m_mesh.DistributionMap(lev),
+                m_mesh.boxArray(lev + 1), amrex::IntVect(2), 1, 0);
+        } else {
+            level_mask.define(
+                m_mesh.boxArray(lev), m_mesh.DistributionMap(lev), 1, 0,
+                amrex::MFInfo());
+            level_mask.setVal(1);
+        }
+
+        const auto& dx = m_mesh.Geom(lev).CellSizeArray();
+        const auto& problo = m_mesh.Geom(lev).ProbLoArray();
+        const amrex::Real cell_vol = dx[0] * dx[1] * dx[2];
+
+        const auto& fld = field(lev);
+        for (amrex::MFIter mfi(fld); mfi.isValid(); ++mfi) {
+            const auto& vbx = mfi.validbox();
+            const auto& field_arr = fld.array(mfi);
+            const auto& mask_arr = level_mask.array(mfi);
+
+            amrex::Real err_fab = 0.0;
+            amrex::LoopOnCpu(vbx, [=, &err_fab](int i, int j, int k) noexcept {
```

Comment:
Looks like this could be `amrex::ReduceSum`

Reply:
done